File: webapp/go/livestream_handler.go
package main

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/jmoiron/sqlx"
	"github.com/labstack/echo-contrib/session"
	"github.com/labstack/echo/v4"
)

type ReserveLivestreamRequest struct {
	Tags        []int64 `json:"tags"`
	Title       string  `json:"title"`
	Description string  `json:"description"`
	// NOTE: コラボ配信の際に便利な自動スケジュールチェック機能
	// DBに記録しないが、コラボレーターがスケジュール的に問題ないか調べて、エラーを返す
	Collaborators []int64 `json:"collaborators"`
	StartAt       int64   `json:"start_at"`
	EndAt         int64   `json:"end_at"`
}

type LivestreamViewerModel struct {
	UserID       int64 `db:"user_id"`
	LivestreamID int64 `db:"livestream_id"`
	CreatedAt    int64 `db:"created_at"`
}

type LivestreamModel struct {
	ID           int64  `db:"id"`
	UserID       int64  `db:"user_id"`
	Title        string `db:"title"`
	Description  string `db:"description"`
	PlaylistUrl  string `db:"playlist_url"`
	ThumbnailUrl string `db:"thumbnail_url"`
	StartAt      int64  `db:"start_at"`
	EndAt        int64  `db:"end_at"`
}

type Livestream struct {
	ID           int64  `json:"id"`
	Owner        User   `json:"owner"`
	Title        string `json:"title"`
	Description  string `json:"description"`
	PlaylistUrl  string `json:"playlist_url"`
	ThumbnailUrl string `json:"thumbnail_url"`
	Tags         []Tag  `json:"tags"`
	StartAt      int64  `json:"start_at"`
	EndAt        int64  `json:"end_at"`
}

type LivestreamTagModel struct {
	ID           int64 `db:"id"`
	LivestreamID int64 `db:"livestream_id"`
	TagID        int64 `db:"tag_id"`
}

type ReservationSlotModel struct {
	ID      int64 `db:"id"`
	Slot    int64 `db:"slot"`
	StartAt int64 `db:"start_at"`
	EndAt   int64 `db:"end_at"`
}

func reserveLivestreamHandler(c echo.Context) error {
	ctx := c.Request().Context()
	defer c.Request().Body.Close()

	if err := verifyUserSession(c); err != nil {
		// echo.NewHTTPErrorが返っているのでそのまま出力
		return err
	}

	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}
	userID, ok := sess.Values[defaultUserIDKey].(int64)
	if !ok {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}

	var req *ReserveLivestreamRequest
	if err := json.NewDecoder(c.Request().Body).Decode(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	// 2024/04/01からの１年間の期間内であるかチェック
	c.Logger().Info("check term")
	var (
		termStartAt    = time.Date(2024, 4, 1, 0, 0, 0, 0, time.UTC)
		termEndAt      = time.Date(2025, 4, 1, 0, 0, 0, 0, time.UTC)
		reserveStartAt = time.Unix(req.StartAt, 0)
		reserveEndAt   = time.Unix(req.EndAt, 0)
	)
	if !(reserveEndAt.Equal(termEndAt) || reserveEndAt.Before(termEndAt)) && (reserveStartAt.Equal(termStartAt) || reserveStartAt.After(termStartAt)) {
		return echo.NewHTTPError(http.StatusBadRequest, "bad reservation time range")
	}

	c.Logger().Info("check collaborators")
	// 各ユーザについて、予約時間帯とかぶるような予約が存在しないか調べる (ある人は同時に複数の配信に物理的に出れない)
	var users []int64
	users = append(users, int64(userID))
	users = append(users, req.Collaborators...)
	for _, user := range users {
		var founds int
		if err := tx.GetContext(ctx, &founds, "SELECT COUNT(*) FROM livestreams WHERE user_id = ? AND  ? >= start_at AND ? <= end_at", user, reserveStartAt, reserveEndAt); err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err)
		}
		if founds >= numReservationSlot {
			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("ユーザ%dが予約できません", user))
		}
	}

	// 予約枠をみて、予約が可能か調べる
	var slots []*ReservationSlotModel
	if err := tx.SelectContext(ctx, &slots, "SELECT * FROM reservation_slots WHERE start_at >= ? AND end_at <= ?", req.StartAt, req.EndAt); err != nil {
		c.Logger().Warnf("予約枠一覧取得でエラー発生: %+v", err)
		return echo.NewHTTPError(http.StatusInternalServerError, err)
	}
	for _, slot := range slots {
		var count int
		if err := tx.GetContext(ctx, &count, "SELECT slot FROM reservation_slots WHERE start_at = ? AND end_at = ?", slot.StartAt, slot.EndAt); err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err)
		}
		c.Logger().Infof("%d ~ %d予約枠の残数 = %d\n", slot.StartAt, slot.EndAt, slot.Slot)
		if count < 1 {
			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("予約区間 %d ~ %dが予約できません", req.StartAt, req.EndAt))
		}
	}

	livestreamModel := &LivestreamModel{
		UserID:      int64(userID),
		Title:       req.Title,
		Description: req.Description,
		// FIXME: プレイリスト、サムネイルは配信環境より配信されるので、それらのURLをPOSTできるようにする
		PlaylistUrl:  "https://d2jpkt808jogxx.cloudfront.net/BigBuckBunny/playlist.m3u8",
		ThumbnailUrl: "https://picsum.photos/200/300",
		StartAt:      req.StartAt,
		EndAt:        req.EndAt,
	}

	c.Logger().Info("insert reservation slot")
	if _, err := tx.ExecContext(ctx, "UPDATE reservation_slots SET slot = slot - 1 WHERE start_at >= ? AND end_at <= ?", req.StartAt, req.EndAt); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	c.Logger().Info("insert livestream")
	rs, err := tx.NamedExecContext(ctx, "INSERT INTO livestreams (user_id, title, description, playlist_url, thumbnail_url, start_at, end_at) VALUES(:user_id, :title, :description, :playlist_url, :thumbnail_url, :start_at, :end_at)", livestreamModel)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	c.Logger().Info("get inserted id")
	livestreamID, err := rs.LastInsertId()
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	livestreamModel.ID = livestreamID

	c.Logger().Info("insert tags")
	// タグ追加
	for _, tagID := range req.Tags {
		if _, err := tx.NamedExecContext(ctx, "INSERT INTO livestream_tags (livestream_id, tag_id) VALUES (:livestream_id, :tag_id)", &LivestreamTagModel{
			LivestreamID: livestreamID,
			TagID:        tagID,
		}); err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}
	}

	livestream, err := fillLivestreamResponse(ctx, tx, *livestreamModel)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusCreated, livestream)
}

func searchLivestreamsHandler(c echo.Context) error {
	ctx := c.Request().Context()
	keyTagName := c.QueryParam("tag")

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	var livestreamModels []*LivestreamModel
	if c.QueryParam("tag") != "" {
		// タグによる取得
		var tagIDList []int
		if err := tx.SelectContext(ctx, &tagIDList, "SELECT id FROM tags WHERE name = ?", keyTagName); err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}

		query, params, err := sqlx.In("SELECT * FROM livestream_tags WHERE id IN (?)", tagIDList)
		if err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}
		var keyTaggedLivestreams []*LivestreamTagModel
		if err := tx.SelectContext(ctx, &keyTaggedLivestreams, query, params...); err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}

		for _, keyTaggedLivestream := range keyTaggedLivestreams {
			ls := LivestreamModel{}
			if err := tx.GetContext(ctx, &ls, "SELECT * FROM livestreams WHERE id = ?", keyTaggedLivestream.LivestreamID); err != nil {
				return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
			}

			livestreamModels = append(livestreamModels, &ls)
		}
	} else {
		// 検索条件なし
		query := `SELECT * FROM livestreams`
		if c.QueryParam("limit") != "" {
			limit, err := strconv.Atoi(c.QueryParam("limit"))
			if err != nil {
				return echo.NewHTTPError(http.StatusBadRequest, err.Error())
			}
			query += fmt.Sprintf(" LIMIT %d", limit)
		}

		if err := tx.SelectContext(ctx, &livestreamModels, query); err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}
	}

	livestreams := make([]Livestream, len(livestreamModels))
	for i := range livestreamModels {
		livestream, err := fillLivestreamResponse(ctx, tx, *livestreamModels[i])
		if err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}
		livestreams[i] = livestream
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, livestreams)
}

func getUserLivestreamsHandler(c echo.Context) error {
	ctx := c.Request().Context()

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}

	userID, ok := sess.Values[defaultUserIDKey].(int64)
	if !ok {
		return echo.NewHTTPError(http.StatusUnauthorized, "failed to find user-id from session")
	}

	var livestreams []*LivestreamModel
	if err := tx.SelectContext(ctx, &livestreams, "SELECT * FROM livestreams WHERE user_id = ?", userID); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, livestreams)
}

// viewerテーブルの廃止
func enterLivestreamHandler(c echo.Context) error {
	ctx := c.Request().Context()
	if err := verifyUserSession(c); err != nil {
		// echo.NewHTTPErrorが返っているのでそのまま出力
		return err
	}

	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}

	userID, ok := sess.Values[defaultUserIDKey].(int64)
	if !ok {
		return echo.NewHTTPError(http.StatusUnauthorized, "failed to find user-id from session")
	}

	livestreamID, err := strconv.Atoi(c.Param("livestream_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	viewer := LivestreamViewerModel{
		UserID:       int64(userID),
		LivestreamID: int64(livestreamID),
		CreatedAt:    time.Now().Unix(),
	}

	if _, err := tx.NamedExecContext(ctx, "INSERT INTO livestream_viewers_history (user_id, livestream_id, created_at) VALUES(:user_id, :livestream_id, :created_at)", viewer); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.NoContent(http.StatusOK)
}

func leaveLivestreamHandler(c echo.Context) error {
	ctx := c.Request().Context()
	if err := verifyUserSession(c); err != nil {
		// echo.NewHTTPErrorが返っているのでそのまま出力
		return err
	}
	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}

	userID, ok := sess.Values[defaultUserIDKey].(int64)
	if !ok {
		return echo.NewHTTPError(http.StatusUnauthorized, "failed to find user-id from session")
	}

	livestreamID, err := strconv.Atoi(c.Param("livestream_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	if _, err := tx.ExecContext(ctx, "DELETE FROM livestream_viewers_history WHERE user_id = ? AND livestream_id = ?", userID, livestreamID); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.NoContent(http.StatusOK)
}

func getLivestreamHandler(c echo.Context) error {
	ctx := c.Request().Context()

	if err := verifyUserSession(c); err != nil {
		return err
	}

	livestreamID, err := strconv.Atoi(c.Param("livestream_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	livestreamModel := LivestreamModel{}
	err = tx.GetContext(ctx, &livestreamModel, "SELECT * FROM livestreams WHERE id = ?", livestreamID)
	if errors.Is(err, sql.ErrNoRows) {
		return echo.NewHTTPError(http.StatusNotFound, err.Error())
	}
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	livestream, err := fillLivestreamResponse(ctx, tx, livestreamModel)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, livestream)
}

func getLivecommentReportsHandler(c echo.Context) error {
	ctx := c.Request().Context()

	if err := verifyUserSession(c); err != nil {
		return err
	}

	livestreamID, err := strconv.Atoi(c.Param("livestream_id"))
	if err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}
	defer tx.Rollback()

	var livestreamModel LivestreamModel
	if err := tx.GetContext(ctx, &livestreamModel, "SELECT * FROM livestreams WHERE id = ?", livestreamID); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	sess, err := session.Get(defaultSessionIDKey, c)
	if err != nil {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}
	userID, ok := sess.Values[defaultUserIDKey].(int64)
	if !ok {
		return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
	}

	if livestreamModel.UserID != userID {
		return echo.NewHTTPError(http.StatusForbidden, "can't get other streamer's livecomment reports")
	}

	var reportModels []*LivecommentReportModel
	if err := tx.SelectContext(ctx, &reportModels, "SELECT * FROM livecomment_reports WHERE livestream_id = ?", livestreamID); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	reports := make([]LivecommentReport, len(reportModels))
	for i := range reportModels {
		report, err := fillLivecommentReportResponse(ctx, tx, *reportModels[i])
		if err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
		}
		reports[i] = report
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
	}

	return c.JSON(http.StatusOK, reports)
}

func fillLivestreamResponse(ctx context.Context, tx *sqlx.Tx, livestreamModel LivestreamModel) (Livestream, error) {
	ownerModel := UserModel{}
	if err := tx.GetContext(ctx, &ownerModel, "SELECT * FROM users WHERE id = ?", livestreamModel.UserID); err != nil {
		return Livestream{}, err
	}
	owner, err := fillUserResponse(ctx, tx, ownerModel)
	if err != nil {
		return Livestream{}, err
	}

	var livestreamTagModels []*LivestreamTagModel
	if err := tx.SelectContext(ctx, &livestreamTagModels, "SELECT * FROM livestream_tags WHERE livestream_id = ?", livestreamModel.ID); err != nil {
		return Livestream{}, err
	}

	tags := make([]Tag, len(livestreamTagModels))
	for i := range livestreamTagModels {
		tagModel := TagModel{}
		if err := tx.GetContext(ctx, &tagModel, "SELECT * FROM tags WHERE id = ?", livestreamTagModels[i].TagID); err != nil {
			return Livestream{}, err
		}

		tags[i] = Tag{
			ID:   tagModel.ID,
			Name: tagModel.Name,
		}
	}

	livestream := Livestream{
		ID:           livestreamModel.ID,
		Owner:        owner,
		Title:        livestreamModel.Title,
		Tags:         tags,
		Description:  livestreamModel.Description,
		PlaylistUrl:  livestreamModel.PlaylistUrl,
		ThumbnailUrl: livestreamModel.ThumbnailUrl,
		StartAt:      livestreamModel.StartAt,
		EndAt:        livestreamModel.EndAt,
	}
	return livestream, nil
}

File: webapp/node/src/handlers/livestream-handler.ts
import { Context } from 'hono'
import { RowDataPacket, ResultSetHeader } from 'mysql2/promise'
import { HonoEnvironment } from '../types/application'
import { verifyUserSessionMiddleware } from '../middlewares/verify-user-session-middleare'
import { defaultUserIDKey } from '../contants'
import {
  LivestreamResponse,
  fillLivestreamResponse,
} from '../utils/fill-livestream-response'
import {
  LivecommentReportResponse,
  fillLivecommentReportResponse,
} from '../utils/fill-livecomment-report-response'
import {
  LivecommentReportsModel,
  LivestreamTagsModel,
  LivestreamsModel,
  ReservationSlotsModel,
  TagsModel,
  UserModel,
} from '../types/models'
import { throwErrorWith } from '../utils/throw-error-with'
import { atoi } from '../utils/integer'

// POST /api/livestream/reservation
export const reserveLivestreamHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/livestream/reservation'>) => {
    const userId = c.get('session').get(defaultUserIDKey) as number // userId is verified by verifyUserSessionMiddleware
    const body = await c.req.json<{
      tags: number[]
      title: string
      description: string
      playlist_url: string
      thumbnail_url: string
      start_at: number
      end_at: number
    }>()

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      // 2023/11/25 10:00からの１年間の期間内であるかチェック
      const termStartAt = Date.UTC(2023, 10, 25, 1)
      const termEndAt = Date.UTC(2024, 10, 25, 1)
      const reserveStartAt = body.start_at * 1000
      const reserveEndAt = body.end_at * 1000

      if (reserveStartAt >= termEndAt || reserveEndAt <= termStartAt) {
        await conn.rollback()
        return c.text('bad reservation time range', 400)
      }

      // 予約枠をみて、予約が可能か調べる
      // NOTE: 並列な予約のoverbooking防止にFOR UPDATEが必要
      const [slots] = await conn
        .query<(ReservationSlotsModel & RowDataPacket)[]>(
          'SELECT * FROM reservation_slots WHERE start_at >= ? AND end_at <= ? FOR UPDATE',
          [body.start_at, body.end_at],
        )
        .catch((error) => {
          console.warn(`予約枠一覧取得でエラー発生: ${error}`)
          return throwErrorWith('failed to get reservation_slots')(error)
        })

      for (const slot of slots) {
        const [[count]] = await conn
          .query<(Pick<ReservationSlotsModel, 'slot'> & RowDataPacket)[]>(
            'SELECT slot FROM reservation_slots WHERE start_at = ? AND end_at = ?',
            [slot.start_at, slot.end_at],
          )
          .catch(throwErrorWith('failed to get reservation_slots'))

        console.info(
          `${slot.start_at} ~ ${slot.end_at} 予約枠の残数 = ${count.slot}`,
        )
        if (count.slot < 1) {
          return c.text(
            `予約期間 ${Math.floor(termStartAt / 1000)} ~ ${Math.floor(
              termEndAt / 1000,
            )}に対して、予約区間 ${body.start_at} ~ ${
              body.end_at
            }が予約できません`,
            400,
          )
        }
      }

      await conn
        .query(
          'UPDATE reservation_slots SET slot = slot - 1 WHERE start_at >= ? AND end_at <= ?',
          [body.start_at, body.end_at],
        )
        .catch(throwErrorWith('failed to update reservation_slot'))
      const [{ insertId: livestreamId }] = await conn
        .query<ResultSetHeader>(
          'INSERT INTO livestreams (user_id, title, description, playlist_url, thumbnail_url, start_at, end_at) VALUES(?, ?, ?, ?, ?, ?, ?)',
          [
            userId,
            body.title,
            body.description,
            body.playlist_url,
            body.thumbnail_url,
            body.start_at,
            body.end_at,
          ],
        )
        .catch(throwErrorWith('failed to insert livestream'))

      // タグ追加
      for (const tagId of body.tags) {
        await conn
          .execute(
            'INSERT INTO livestream_tags (livestream_id, tag_id) VALUES (?, ?)',
            [livestreamId, tagId],
          )
          .catch(throwErrorWith('failed to insert livestream tag'))
      }

      const response = await fillLivestreamResponse(
        conn,
        {
          id: livestreamId,
          user_id: userId,
          title: body.title,
          description: body.description,
          playlist_url: body.playlist_url,
          thumbnail_url: body.thumbnail_url,
          start_at: body.start_at,
          end_at: body.end_at,
        },
        c.get('runtime').fallbackUserIcon,
      ).catch(throwErrorWith('failed to fill livestream'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json(response, 201)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// GET /api/livestream/search
export const searchLivestreamsHandler = async (
  c: Context<HonoEnvironment, '/api/livestream/search'>,
) => {
  const keyTagName = c.req.query('tag')

  const conn = await c.get('pool').getConnection()
  await conn.beginTransaction()

  try {
    const livestreams: (LivestreamsModel & RowDataPacket)[] = []

    if (keyTagName) {
      // タグによる取得
      const [tagIds] = await conn
        .query<(Pick<TagsModel, 'id'> & RowDataPacket)[]>(
          'SELECT id FROM tags WHERE name = ?',
          [keyTagName],
        )
        .catch(throwErrorWith('failed to get tag'))

      const [livestreamTags] = await conn
        .query<(LivestreamTagsModel & RowDataPacket)[]>(
          'SELECT * FROM livestream_tags WHERE tag_id IN (?) ORDER BY livestream_id DESC',
          [tagIds.map((tag) => tag.id)],
        )
        .catch(throwErrorWith('failed to get keyTaggedLivestreams'))

      for (const livestreamTag of livestreamTags) {
        const [[livestream]] = await conn
          .query<(LivestreamsModel & RowDataPacket)[]>(
            'SELECT * FROM livestreams WHERE id = ?',
            [livestreamTag.livestream_id],
          )
          .catch(throwErrorWith('failed to get livestreams'))

        livestreams.push(livestream)
      }
    } else {
      // 検索条件なし
      let query = `SELECT * FROM livestreams ORDER BY id DESC`
      const limit = c.req.query('limit')
      if (limit) {
        const limitNumber = atoi(limit)
        if (limitNumber === false) {
          return c.text('limit query parameter must be integer', 400)
        }
        query += ` LIMIT ${limitNumber}`
      }

      const [results] = await conn
        .query<(LivestreamsModel & RowDataPacket)[]>(query)
        .catch(throwErrorWith('failed to get livestreams'))

      livestreams.push(...results)
    }

    const livestreamResponses: LivestreamResponse[] = []
    for (const livestream of livestreams) {
      const livestreamResponse = await fillLivestreamResponse(
        conn,
        livestream,
        c.get('runtime').fallbackUserIcon,
      ).catch(throwErrorWith('failed to fill livestream'))
      livestreamResponses.push(livestreamResponse)
    }

    await conn.commit().catch(throwErrorWith('failed to commit'))

    return c.json(livestreamResponses)
  } catch (error) {
    await conn.rollback()
    return c.text(`Internal Server Error\n${error}`, 500)
  } finally {
    await conn.rollback()
    conn.release()
  }
}

// GET /api/livestream
export const getMyLivestreamsHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/livestream'>) => {
    const userId = c.get('session').get(defaultUserIDKey) as number // userId is verified by verifyUserSessionMiddleware

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      const [livestreams] = await conn
        .query<(LivestreamsModel & RowDataPacket)[]>(
          'SELECT * FROM livestreams WHERE user_id = ?',
          [userId],
        )
        .catch(throwErrorWith('failed to get livestreams'))

      const livestreamResponses: LivestreamResponse[] = []
      for (const livestream of livestreams) {
        const livestreamResponse = await fillLivestreamResponse(
          conn,
          livestream,
          c.get('runtime').fallbackUserIcon,
        ).catch(throwErrorWith('failed to fill livestream'))

        livestreamResponses.push(livestreamResponse)
      }

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json(livestreamResponses)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// GET /api/user/:username/livestream
export const getUserLivestreamsHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/user/:username/livestream'>) => {
    const username = c.req.param('username')

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      const [[user]] = await conn
        .query<(UserModel & RowDataPacket)[]>(
          'SELECT * FROM users WHERE name = ?',
          [username],
        )
        .catch(throwErrorWith('failed to get user'))

      if (!user) {
        return c.text('user not found', 404)
      }

      const [livestreams] = await conn
        .query<(LivestreamsModel & RowDataPacket)[]>(
          'SELECT * FROM livestreams WHERE user_id = ?',
          [user.id],
        )
        .catch(throwErrorWith('failed to get livestreams'))

      const livestreamResponses: LivestreamResponse[] = []
      for (const livestream of livestreams) {
        const livestreamResponse = await fillLivestreamResponse(
          conn,
          livestream,
          c.get('runtime').fallbackUserIcon,
        ).catch(throwErrorWith('failed to fill livestream'))

        livestreamResponses.push(livestreamResponse)
      }

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json(livestreamResponses)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// POST /api/livestream/:livestream_id/enter
export const enterLivestreamHandler = [
  verifyUserSessionMiddleware,
  async (
    c: Context<HonoEnvironment, '/api/livestream/:livestream_id/enter'>,
  ) => {
    const userId = c.get('session').get(defaultUserIDKey) as number // userId is verified by verifyUserSessionMiddleware
    const livestreamId = atoi(c.req.param('livestream_id'))
    if (livestreamId === false) {
      return c.text('livestream_id in path must be integer', 400)
    }

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      await conn
        .query(
          'INSERT INTO livestream_viewers_history (user_id, livestream_id, created_at) VALUES(?, ?, ?)',
          [userId, livestreamId, Date.now()],
        )
        .catch(throwErrorWith('failed to insert livestream_view_history'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      // eslint-disable-next-line unicorn/no-null
      return c.body(null)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// DELETE /api/livestream/:livestream_id/exit
export const exitLivestreamHandler = [
  verifyUserSessionMiddleware,
  async (
    c: Context<HonoEnvironment, '/api/livestream/:livestream_id/exit'>,
  ) => {
    const userId = c.get('session').get(defaultUserIDKey) as number // userId is verified by verifyUserSessionMiddleware
    const livestreamId = atoi(c.req.param('livestream_id'))
    if (livestreamId === false) {
      return c.text('livestream_id in path must be integer', 400)
    }

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      await conn
        .query(
          'DELETE FROM livestream_viewers_history WHERE user_id = ? AND livestream_id = ?',
          [userId, livestreamId],
        )
        .catch(throwErrorWith('failed to delete livestream_view_history'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      // eslint-disable-next-line unicorn/no-null
      return c.body(null)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// GET /api/livestream/:livestream_id
export const getLivestreamHandler = [
  verifyUserSessionMiddleware,
  async (c: Context<HonoEnvironment, '/api/livestream/:livestream_id'>) => {
    const livestreamId = atoi(c.req.param('livestream_id'))
    if (livestreamId === false) {
      return c.text('livestream_id in path must be integer', 400)
    }

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()

    try {
      const [[livestream]] = await conn
        .query<(LivestreamsModel & RowDataPacket)[]>(
          'SELECT * FROM livestreams WHERE id = ?',
          [livestreamId],
        )
        .catch(throwErrorWith('failed to get livestream'))

      if (!livestream) {
        return c.text('not found livestream that has the given id', 404)
      }

      const livestreamResponse = await fillLivestreamResponse(
        conn,
        livestream,
        c.get('runtime').fallbackUserIcon,
      ).catch(throwErrorWith('failed to fill livestream'))

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json(livestreamResponse)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]

// GET /api/livestream/:livestream_id/report
export const getLivecommentReportsHandler = [
  verifyUserSessionMiddleware,
  async (
    c: Context<HonoEnvironment, '/api/livestream/:livestream_id/report'>,
  ) => {
    const userId = c.get('session').get(defaultUserIDKey) as number // userId is verified by verifyUserSessionMiddleware
    const livestreamId = atoi(c.req.param('livestream_id'))
    if (livestreamId === false) {
      return c.text('livestream_id in path must be integer', 400)
    }

    const conn = await c.get('pool').getConnection()
    await conn.beginTransaction()
    try {
      const [[livestream]] = await conn
        .query<(LivestreamsModel & RowDataPacket)[]>(
          'SELECT * FROM livestreams WHERE id = ?',
          [livestreamId],
        )
        .catch(throwErrorWith('failed to get livestream'))

      if (livestream.user_id !== userId) {
        return c.text("can't get other streamer's livecomment reports", 403)
      }

      const [livecommentReports] = await conn
        .query<(LivecommentReportsModel & RowDataPacket)[]>(
          'SELECT * FROM livecomment_reports WHERE livestream_id = ?',
          [livestreamId],
        )
        .catch(throwErrorWith('failed to get livecomment reports'))

      const reportResponses: LivecommentReportResponse[] = []
      for (const livecommentReport of livecommentReports) {
        const report = await fillLivecommentReportResponse(
          conn,
          livecommentReport,
          c.get('runtime').fallbackUserIcon,
        ).catch(throwErrorWith('failed to fill livecomment report'))
        reportResponses.push(report)
      }

      await conn.commit().catch(throwErrorWith('failed to commit'))

      return c.json(reportResponses)
    } catch (error) {
      await conn.rollback()
      return c.text(`Internal Server Error\n${error}`, 500)
    } finally {
      await conn.rollback()
      conn.release()
    }
  },
]
